spec/Section 2 -- Language.md

```diff
@@ -47,31 +47,31 @@ match, however some lookahead restrictions include additional constraints.
 ## Source Text
 
 SourceCharacter ::
-  - "U+0009"
-  - "U+000A"
-  - "U+000D"
-  - "U+0020–U+FFFF"
+  - "Any Unicode scalar value"
 
-GraphQL documents are expressed as a sequence of
-[Unicode](https://unicode.org/standard/standard.html) code points (informally
-referred to as *"characters"* through most of this specification). However, with
-few exceptions, most of GraphQL is expressed only in the original non-control
-ASCII range so as to be as widely compatible with as many existing tools,
-languages, and serialization formats as possible and avoid display issues in
-text editors and source control.
+GraphQL documents are interpreted from a source text, which is a sequence of
+{SourceCharacter}, each {SourceCharacter} being a
+[Unicode scalar value](https://www.unicode.org/glossary/#unicode_scalar_value)
+which may be any Unicode code point from U+0000 to U+D7FF or U+E000 to U+10FFFF
+(informally referred to as *"characters"* through most of this specification).
 
-Note: Non-ASCII Unicode characters may appear freely within {StringValue} and
-{Comment} portions of GraphQL.
+GraphQL documents may be stored and transmitted in any encoding. Regardless of
+any implementation-specific encoding, a GraphQL document's source text should be
+interpreted as if it was an sequence of {SourceCharacter} Unicode scalar values.
 
+As an example, implementations which use [UTF-16](https://www.unicode.org/glossary/#UTF_16)
+to store GraphQL documents (for example, JavaScript or Java) may natively encode
+[supplementary code points](https://www.unicode.org/glossary/#supplementary_code_point)
+(those above U+FFFF) as a [surrogate pair](https://www.unicode.org/glossary/#surrogate_pair).
+Surrogate code points are not Unicode scalar values, but it is compliant to
+interpret such surrogate pairs as single scalar values within the sequence of
+{SourceCharacter}. However individual surrogate code points must not be
+contained within this sequence.
```

Comment:
@andimarek this paragraph is based on our discussion around abstracting away an implementation specific encoding which may itself contain surrogate pairs

Reply:
I find this confusing to read and understand tbh. I will try to unpcack my thinking:

It is true that JS/Java use UTF-16 internally to store Strings. But this is purely a runtime, in memory representation.

If I read a Document via IO (http request, file) I need to assume an encoding of this document to interpret the bytes I get. Lets say it is UTF-8.

UTF-8 encoded Document -> Internal JS UTF-16 representation of the document as String -> GraphQL Parser/Lexer -> Internal JS UTF-16 representation of evaluated Strings

This means the GraphQL Lexer always operates on JS UTF-16 [code units](https://unicode.org/glossary/#code_unit) and code points outside of the BMP are always representated as surrogate pairs in memory (for JS/Java at least).

Are you trying to say that this internal UTF-16 representation of documents is acceptable and doesn't conflict with the definition of `SourceCharacter`?

Reply:
I'm definitely open to alternative wordings. What I wrote makes sense to me of course, but whatever makes this more clear I'm definitely open to. Let me know what you think.

> Are you trying to say that this internal UTF-16 representation of documents is acceptable and doesn't conflict with the definition of SourceCharacter?

Correct. For example, in saying that GraphQL accepts any Unicode Scalar Value it should be valid to provide U+1F600 as a Source Character. However if your implementation uses UTF-16, then 0x1F600 does not fit in 16 bits, so UTF-16 defines that this code point should be represented by the code unit pair 0xD800, 0xDC00, and thus these two code units will appear in the input for a GraphQL source text. These two code units do not define Unicode Scalar Values since they are surrogates, so then does this result in an input that is no longer valid GraphQL source text? That can't be the case since we claim that both a source text is a sequence of "any Unicode Scalar Value" *and* "GraphQL documents may be stored and transmitted in any encoding." If both of these are true, then we must accept such an input. The way we accept it is that a GraphQL source text "should be
interpreted as if it was an sequence of {SourceCharacter} Unicode scalar values" and the UTF-16 specific implementation detail of a surrogate pair (0xD800, 0xDC00) encodes one Unicode scalar value (U+1F600) in that sequence.

Reply:
My feeling is that we are making it a bit more complicated than it is: 

Maybe we could simply highlight that `SourceCharacter` is a "logical or semantic view" indepdended of the actual encoding used in the implementation, because they are code points and not code units with a link to the Unicode spec.

Reply:
I'm definitely okay with simplifying the language, and what you're suggesting here is definitely accurate. My concern was that we would need this guidance for implementers, but maybe the reference implementation is enough (would love your thoughts on whether the documentation within that new PR is clear enough - https://github.com/graphql/graphql-js/pull/3117/)

Reply:
I would recommend to simplify this here and depending on the feedback we get we can expand it again. There reference impl. is probably a better place for the details as you pointed out.